File: ios/ConverseNotificationExtension/Xmtp/Messages.swift
//
//  Messages.swift
//  ConverseNotificationExtension
//
//  Created by Noe Malzieu on 28/08/2023.
//

import Foundation
import XMTP
import Alamofire
import Intents


func handleV3Welcome(xmtpClient: XMTP.Client, apiURI: String?, pushToken: String?, conversation: XMTP.Conversation, welcomeTopic: String, bestAttemptContent: inout UNMutableNotificationContent) async -> (shouldShowNotification: Bool, messageId: String?) {
  var shouldShowNotification = false
  let messageId = "welcome-" + conversation.topic
  do {
    // group is already synced in getNewGroup method
    let spamScore = await computeSpamScoreV3Welcome(client: xmtpClient, conversation: conversation, apiURI: apiURI)
    if spamScore < 0 { // Message is going to main inbox
      // Consent list is loaded in computeSpamScoreGroupWelcome
      let convoAllowed = try await xmtpClient.preferences.conversationState(conversationId: conversation.id) == .allowed
      let convoDenied = try await xmtpClient.preferences.conversationState(conversationId: conversation.id) == .denied
      // If group is already consented (either way) then don't show a notification for welcome as this will likely be a second+ installation
      if !convoAllowed && !convoDenied {
        
        shouldShowNotification = true
        if case .group(let group) = conversation {
          let groupName = try group.groupName()
          bestAttemptContent.title = groupName
          bestAttemptContent.body = "You have been added to a new group"
        } else if case .dm(let dm) = conversation {
          bestAttemptContent.title = "New DM"
          bestAttemptContent.body = "You have a new direct message"
        }

      } else {
        shouldShowNotification = false
      }
    } else if spamScore == 0 { // Message is Request
      shouldShowNotification = false
      trackNewRequest()
    } else { // Message is Spam
      shouldShowNotification = false
    }
    
  } catch {
    sentryTrackError(error: error, extras: ["message": "NOTIFICATION_SAVE_MESSAGE_ERROR_3", "topic": conversation.topic])
    print("[NotificationExtension] Error handling group invites: \(error)")
  }
  
  return (shouldShowNotification, messageId)
}

func handleV3Message(xmtpClient: XMTP.Client, envelope: XMTP.Xmtp_MessageApi_V1_Envelope, apiURI: String?, bestAttemptContent: inout UNMutableNotificationContent) async -> (shouldShowNotification: Bool, messageId: String?, messageIntent: INSendMessageIntent?) {
  var shouldShowNotification = false
  let contentTopic = envelope.contentTopic
  var messageId: String? = nil
  var messageIntent: INSendMessageIntent? = nil
  
  do {
    
    if let conversation = try await xmtpClient.findConversationByTopic(topic: envelope.contentTopic) {
      try await conversation.sync()
      if var decodedMessage = try? await decodeMessage(xmtpClient: xmtpClient, envelope: envelope) {
        // For now, use the group member linked address as "senderAddress"
        // @todo => make inboxId a first class citizen
        if let senderAddresses = try await conversation.members().first(where: {$0.inboxId == decodedMessage.senderInboxId})?.addresses {
          // This is a bit yucky since we are doing some gymnastics around inboxId vs address.
          // Will have a fast follow if we see any issues with push
          decodedMessage.senderInboxId = senderAddresses[0]
        }

        let decodedMessageResult = handleMessageByContentType(decodedMessage: decodedMessage, xmtpClient: xmtpClient);
        messageId = decodedMessageResult.id
        if decodedMessageResult.senderAddress?.lowercased() == xmtpClient.inboxID.lowercased() || decodedMessageResult.senderAddress?.lowercased() == xmtpClient.address.lowercased() || decodedMessageResult.forceIgnore {
        } else {
          let spamScore = await computeSpamScoreV3Message(client: xmtpClient, conversation: conversation, decodedMessage: decodedMessage, apiURI: apiURI)
          
          if spamScore < 0 { // Message is going to main inbox
            shouldShowNotification = true
            if case .group(let group) = conversation {
              if let groupName = try? group.groupName() {
                bestAttemptContent.title = groupName
              }
              // We replaced decodedMessage.senderAddress from inboxId to actual address
              // so it appears well in the app until inboxId is a first class citizen
              if let senderProfileSocials = await getProfile(account: xmtpClient.address, address: decodedMessage.senderInboxId) {
                bestAttemptContent.subtitle = getPreferredName(address: decodedMessage.senderInboxId, socials: senderProfileSocials)
              }

              if let content = decodedMessageResult.content {
                bestAttemptContent.body = content
              }
              let groupImage = try? group.groupImageUrlSquare()
              messageIntent = getIncomingGroupMessageIntent(group: group, content: bestAttemptContent.body, senderId: decodedMessage.senderInboxId, senderName: bestAttemptContent.subtitle)
            } else if case .dm(let dm) = conversation {
              var senderAvatar: String? = nil
              if let senderProfileSocials = await getProfile(account: xmtpClient.address, address: decodedMessage.senderInboxId) {
                bestAttemptContent.title = getPreferredName(address: decodedMessage.senderInboxId, socials: senderProfileSocials)
                senderAvatar = getPreferredAvatar(socials: senderProfileSocials)
              }
              if let content = decodedMessageResult.content {
                bestAttemptContent.body = content
              }
              messageIntent = getIncoming1v1MessageIntent(
                topic: contentTopic,
                senderId: decodedMessage.senderInboxId,
                senderName: bestAttemptContent.title,
                senderAvatar: senderAvatar, content: bestAttemptContent.body
              )
            }
            
          } else if spamScore == 0 { // Message is Request
            shouldShowNotification = false
          } else { // Message is Spam
            shouldShowNotification = false
          }
        }
      }
    }
  } catch {
    sentryTrackError(error: error, extras: ["message": "NOTIFICATION_SAVE_MESSAGE_ERROR_4", "topic": contentTopic])
    print("[NotificationExtension] Error handling group message: \(error)")
    
  }
  return (shouldShowNotification, messageId, messageIntent)
}

func loadSavedMessages() -> [SavedNotificationMessage] {
  let mmkv = getMmkv()
  let savedMessagesString = mmkv?.string(forKey: "saved-notifications-messages")
  if (savedMessagesString == nil) {
    return []
  } else {
    let decoder = JSONDecoder()
    do {
      let decoded = try decoder.decode([SavedNotificationMessage].self, from: savedMessagesString!.data(using: .utf8)!)
      return decoded
    } catch {
      return []
    }
  }
}

func decodeMessage(xmtpClient: XMTP.Client, envelope: XMTP.Xmtp_MessageApi_V1_Envelope) async throws -> DecodedMessage? {
  // If topic is MLS, the conversation should already be there
  // @todo except if it's new convo => call sync before?
  if (isV3MessageTopic(topic: envelope.contentTopic)) {
    if let conversation = try! await xmtpClient.findConversationByTopic(topic: envelope.contentTopic) {
      do {
        sentryAddBreadcrumb(message: "[NotificationExtension] Syncing Group")
        try await conversation.sync()
        sentryAddBreadcrumb(message: "[NotificationExtension] Decoding group message...")
        let messageBytes = envelope.message
        let message = try await conversation.processMessage(messageBytes: messageBytes)
        let decodedMessage = try message.decode()
        sentryAddBreadcrumb(message:"[NotificationExtension] Group message decoded!")
        return decodedMessage
      } catch {
        sentryTrackMessage(message: "NOTIFICATION_DECODING_ERROR", extras: ["error": error, "envelope": envelope])
        sentryAddBreadcrumb(message: "[NotificationExtension] ERROR WHILE DECODING \(error)")
        return nil
      }
    } else {
      sentryTrackMessage(message: "NOTIFICATION_GROUP_NOT_FOUND", extras: ["envelope": envelope])
      return nil
    }
  }
  return nil
}

func handleMessageByContentType(decodedMessage: DecodedMessage, xmtpClient: XMTP.Client) -> (content: String?, senderAddress: String?, forceIgnore: Bool, id: String?) {
  var contentType = getContentTypeString(type: decodedMessage.encodedContent.type)
  var contentToReturn: String?
  var contentToSave: String?
  var referencedMessageId: String?
  var forceIgnore = false
  
  var messageContent = try? decodedMessage.content() as Any
  
  if (contentType.starts(with: "xmtp.org/reply:")) {
    let replyContent = messageContent as? Reply
    if let reply = replyContent {
      referencedMessageId = reply.reference
      contentType = getContentTypeString(type: reply.contentType)
      messageContent = reply.content
    }
    
  }
  
  do {
    switch contentType {
      
    case let type where type.starts(with: "xmtp.org/text:"):
      contentToSave = messageContent as? String
      contentToReturn = contentToSave
      
    case let type where type.starts(with: "xmtp.org/remoteStaticAttachment:"):
      let remoteAttachment = messageContent as! RemoteAttachment
      contentToSave = getJsonRemoteAttachment(remoteAttachment: remoteAttachment)
      contentToReturn = "📎 Media"
      
    case let type where type.starts(with: "xmtp.org/transactionReference:") ||
      type.starts(with: "coinbase.com/coinbase-messaging-payment-activity:"):
      contentToSave = messageContent as? String
      contentToReturn = "💸 Transaction"
      
    case let type where type.starts(with: "xmtp.org/reaction:"):
      let reaction = messageContent as? Reaction
      let action = reaction?.action.rawValue
      let schema = reaction?.schema.rawValue
      let content = reaction?.content
      referencedMessageId = reaction?.reference
      
      if action == "removed" {
        forceIgnore = true
      } else if action != "removed" && schema == "unicode", let reactionContent = content {
        contentToReturn = "Reacted \(reactionContent) to a message"
      } else {
        contentToReturn = "Reacted to a message"
      }
      
      // For groups: notify reactions to messages from me only
      if (isV3MessageTopic(topic: decodedMessage.topic) && referencedMessageId != nil) {
        forceIgnore = !(try isGroupMessageFromMe(xmtpClient: xmtpClient, messageId: referencedMessageId!))
      }
      
      
      if let validReaction = reaction {
        contentToSave = getJsonReaction(reaction: validReaction)
      } else {
        contentToSave = nil
      }
      
    case let type where type.starts(with: "xmtp.org/readReceipt:"):
      contentToSave = nil
      
    default:
      sentryTrackMessage(message: "NOTIFICATION_UNKNOWN_CONTENT_TYPE", extras: ["contentType": contentType, "topic": decodedMessage.topic])
      print("[NotificationExtension] UNKOWN CONTENT TYPE: \(contentType)")
      return (nil, decodedMessage.senderInboxId, false, nil)
    }
    
    if (isDebugAccount(account: xmtpClient.address)) {
      sentryAddBreadcrumb(message: "Finished handling message content - \(contentToReturn ?? "EMPTY") - tosave \(contentToSave ?? "EMPTY")")
    }

    return (contentToReturn, decodedMessage.senderInboxId, forceIgnore, decodedMessage.id)
  } catch {
    let errorType = contentType.split(separator: "/").last ?? "UNKNOWN"
    sentryTrackError(error: error, extras: ["message": "NOTIFICATION_\(errorType)_ERROR", "topic": decodedMessage.topic])
    return (nil, nil, false, nil)
  }
}

func getContentTypeString(type: ContentTypeID) -> String {
  return "\(type.authorityID)/\(type.typeID):\(type.versionMajor).\(type.versionMinor)"
}

func getJsonRemoteAttachment(remoteAttachment: RemoteAttachment) -> String? {
  do {
    let dictionary = NSDictionary(dictionary: ["url": remoteAttachment.url, "contentDigest": remoteAttachment.contentDigest, "secret": remoteAttachment.secret.base64EncodedString(), "salt": remoteAttachment.salt.base64EncodedString(), "nonce": remoteAttachment.nonce.base64EncodedString(), "scheme": remoteAttachment.scheme.rawValue, "contentLength": remoteAttachment.contentLength ?? 0, "filename": remoteAttachment.filename ?? ""])
    let jsonData = try JSONSerialization.data(withJSONObject: dictionary, options: [])
    let jsonString = String(data: jsonData, encoding: .utf8)
    return jsonString
  } catch {
    print("Error converting dictionary to JSON string: \(error.localizedDescription)")
    return nil
  }
}

func getJsonReaction(reaction: Reaction) -> String? {
  do {
    let reference = reaction.reference;
    let schema = reaction.schema.rawValue;
    let action = reaction.action.rawValue;
    let content = reaction.content;
    let dictionary = NSDictionary(dictionary: ["reference": reference, "action": action, "content": content, "schema": schema])
    let jsonData = try JSONSerialization.data(withJSONObject: dictionary, options: [])
    let jsonString = String(data: jsonData, encoding: .utf8)
    return jsonString
  } catch {
    print("Error converting dictionary to JSON string: \(error.localizedDescription)")
    return nil
  }
}

func isGroupMessageFromMe(xmtpClient: Client, messageId: String) throws -> Bool {
  if let message = try xmtpClient.findMessage(messageId: messageId) {
    return message.decodeOrNull()?.senderInboxId == xmtpClient.inboxID
  } else {
    return false
  }
}

File: android/app/src/main/java/com/converse/xmtp/Messages.kt
package com.converse.xmtp

import android.content.Context
import android.util.Log
import com.beust.klaxon.Klaxon

import com.converse.*
import com.converse.PushNotificationsService.Companion.TAG
import com.google.firebase.messaging.RemoteMessage
import computeSpamScoreDmWelcome
import computeSpamScoreGroupMessage
import computeSpamScoreGroupWelcome
import isConversationAllowed
import isConversationBlocked

import org.json.JSONObject
import org.xmtp.android.library.Client
import org.xmtp.android.library.Conversation
import org.xmtp.android.library.DecodedMessage
import org.xmtp.android.library.codecs.Reaction
import org.xmtp.android.library.codecs.RemoteAttachment
import org.xmtp.android.library.codecs.Reply
import org.xmtp.proto.message.api.v1.MessageApiOuterClass

import org.xmtp.proto.message.contents.Content


data class NotificationDataResult(
    val title: String = "",
    val subtitle: String? = null,
    val body: String = "",
    val remoteMessage: RemoteMessage? = null,
    val messageId: String? = null,
    val shouldShowNotification: Boolean = false,
    val avatar: String? = null,
    val isGroup: Boolean = false
)

data class DecodedMessageResult(
    val content: String? = null,
    val senderAddress: String? = null,
    val forceIgnore: Boolean = false,
    val id: String? = null
)

suspend fun handleV3Message(
    appContext: Context,
    xmtpClient: Client,
    envelope: MessageApiOuterClass.Envelope,
    remoteMessage: RemoteMessage,
): NotificationDataResult {
    val conversation = xmtpClient.findConversationByTopic(envelope.contentTopic)
    if (conversation == null) {
        Log.d("PushNotificationsService", "No conversation found for ${envelope.contentTopic}")
        return NotificationDataResult()
    }
    sentryTrackMessage(
        "[NotificationExtension] Found conversation",
        mapOf()
    )

    conversation.sync()

    sentryTrackMessage(
        "[NotificationExtension] Done syncing group",
        mapOf()
    )

    val decodedMessage = conversation.processMessage(envelope.message.toByteArray()).decode()
    when (conversation) {
        is Conversation.Group -> {
            // Handle the Group case
            return handleGroupMessage(appContext, conversation, decodedMessage, xmtpClient, remoteMessage)
        }
        is Conversation.Dm -> {
            // Handle the Dm case
            return handleDmMessage(appContext, conversation, decodedMessage, xmtpClient, remoteMessage)
        }
    }
}

suspend fun handleDmMessage(
    appContext: Context,
    conversation: Conversation.Dm,
    decodedMessage: DecodedMessage,
    xmtpClient: Client,
    remoteMessage: RemoteMessage
): NotificationDataResult {
// For now, use the conversation member linked address as "senderAddress"
    val dm = conversation.dm
    // @todo => make inboxId a first class citizen
    dm.members().firstOrNull { it.inboxId == decodedMessage.senderAddress }?.addresses?.get(0)?.let { senderAddress ->
        decodedMessage.senderAddress = senderAddress
    }
    var conversationTitle = ""

    val decodedMessageResult = handleMessageByContentType(
        appContext,
        decodedMessage,
        xmtpClient,
    )

    decodedMessageResult.senderAddress?.let {
        conversationTitle = shortAddress(it)
    }

    val senderProfile =
        decodedMessageResult.senderAddress?.let { getProfile(appContext, xmtpClient.address, it) }
    var senderAvatar: String? = null
    senderProfile?.let { senderProfile ->
        conversationTitle = getPreferredName(address = decodedMessageResult.senderAddress, socials = senderProfile.socials)
        senderAvatar = getPreferredAvatar(socials = senderProfile.socials)
    }

    val shouldShowNotification = if (decodedMessageResult.senderAddress != xmtpClient.address && !decodedMessageResult.forceIgnore && decodedMessageResult.content != null) {
        if ((conversationTitle == null || conversationTitle!!.isEmpty()) && decodedMessageResult.senderAddress != null) {
            conversationTitle = shortAddress(decodedMessageResult.senderAddress)
        }
        true
    } else {
        Log.d(TAG, "[NotificationExtension] Not showing a notification")
        false
    }

    return NotificationDataResult(
        title = conversationTitle,
        body = decodedMessageResult.content ?: "",
        remoteMessage = remoteMessage,
        messageId = decodedMessageResult.id,
        shouldShowNotification = shouldShowNotification,
        avatar = senderAvatar
    )
}

suspend fun handleGroupMessage(
    appContext: Context,
    convoGroup: Conversation.Group,
    decodedMessage: DecodedMessage,
    xmtpClient: Client,
    remoteMessage: RemoteMessage
): NotificationDataResult {
// For now, use the conversation member linked address as "senderAddress"
    val group = convoGroup.group
    // @todo => make inboxId a first class citizen
    group.members().firstOrNull { it.inboxId == decodedMessage.senderAddress }?.addresses?.get(0)?.let { senderAddress ->
        decodedMessage.senderAddress = senderAddress
    }
    val decodedMessageResult = handleMessageByContentType(
        appContext,
        decodedMessage,
        xmtpClient,
    )

    var shouldShowNotification = if (decodedMessageResult.senderAddress?.lowercase() != xmtpClient.inboxId.lowercase() && decodedMessageResult.senderAddress?.lowercase() != xmtpClient.address.lowercase() && !decodedMessageResult.forceIgnore && decodedMessageResult.content != null) {
        true
    } else {
        Log.d(TAG, "[NotificationExtension] Not showing a notification")
        false
    }

    var subtitle: String? = null

    if (shouldShowNotification) {
        val mmkv = getMmkv(appContext)
        var apiURI = mmkv?.decodeString("api-uri")
        if (apiURI == null) {
            apiURI = getAsyncStorage("api-uri")
        }
        val spamScore = computeSpamScoreGroupMessage(xmtpClient, group, decodedMessage, apiURI)
        if (spamScore < 0) {
            // Message is going to main inbox
            // We replaced decodedMessage.senderAddress from inboxId to actual address
            // so it appears well in the app until inboxId is a first class citizen
            val senderProfile = getProfile(appContext, xmtpClient.address, decodedMessage.senderAddress)
            senderProfile?.let { senderProfile ->
                subtitle = getPreferredName(decodedMessage.senderAddress, senderProfile.socials)
            }
        } else if (spamScore == 0) { // Message is Request
          shouldShowNotification = false
        } else { // Message is Spam
          shouldShowNotification = false
        }
    }

    return NotificationDataResult(
        title = group.name,
        subtitle = subtitle,
        body = decodedMessageResult.content ?: "",
        remoteMessage = remoteMessage,
        messageId = decodedMessageResult.id,
        shouldShowNotification = shouldShowNotification,
        isGroup = true,
        avatar = group.imageUrlSquare
    )
}

fun handleMessageByContentType(
    appContext: Context,
    decodedMessage: DecodedMessage,
    xmtpClient: Client,
): DecodedMessageResult {
    var contentType = getContentTypeString(decodedMessage.encodedContent.type)
    var contentToReturn: String?
    var contentToSave: String?
    var referencedMessageId: String? = null
    var forceIgnore = false

    var messageContent = decodedMessage.content<Any>()

    if (contentType.startsWith("xmtp.org/reply:")) {
        val replyContent = messageContent as Reply
        referencedMessageId = replyContent.reference
        contentType = getContentTypeString(replyContent.contentType)
        messageContent = replyContent.content
    }

    try {
        when {
            contentType.startsWith("xmtp.org/text:") -> {
                contentToSave = messageContent as String
                contentToReturn = contentToSave
            }

            contentType.startsWith("xmtp.org/remoteStaticAttachment:") -> {
                val remoteAttachment = messageContent as RemoteAttachment
                contentToSave = getJsonRemoteAttachment(remoteAttachment)
                contentToReturn = "📎 Media"
            }

            contentType.startsWith("xmtp.org/transactionReference:") || contentType.startsWith("coinbase.com/coinbase-messaging-payment-activity:") -> {
                contentToSave = messageContent as String
                contentToReturn = "💸 Transaction"
            }

            contentType.startsWith("xmtp.org/reaction:") -> {
                val reaction = messageContent as Reaction?
                val action = reaction?.action?.javaClass?.simpleName?.lowercase()
                val schema = reaction?.schema?.javaClass?.simpleName?.lowercase()
                val content = reaction?.content
                referencedMessageId = reaction?.reference
                forceIgnore = action == "removed" || (isV3MessageTopic(decodedMessage.topic) && referencedMessageId !== null && !isGroupMessageFromMe(xmtpClient, referencedMessageId))
                contentToSave = getJsonReaction(decodedMessage)
                contentToReturn = when {
                    action != "removed" && schema == "unicode" && content != null -> "Reacted $content to a message"
                    else -> "Reacted to a message"
                }
            }

            contentType.startsWith("xmtp.org/readReceipt:") -> {
                // Ignoring
                contentToSave = null
                contentToReturn = null
            }

            else -> {
                sentryTrackMessage(
                    "NOTIFICATION_UNKNOWN_CONTENT_TYPE",
                    mapOf("contentType" to contentType, "topic" to decodedMessage.topic)
                )
                println("[NotificationExtension] UNKNOWN CONTENT TYPE: $contentType")
                return DecodedMessageResult(null, decodedMessage.senderAddress, false, null)
            }
        }

        // Save message to mmkv
        contentToSave?.let {
            saveMessageToStorage(
                appContext = appContext,
                account = xmtpClient.address,
                decodedMessage = decodedMessage,
                content = it,
                contentType = contentType,
                referencedMessageId = referencedMessageId
            )
        }

        return DecodedMessageResult(contentToReturn, decodedMessage.senderAddress, forceIgnore, decodedMessage.id)
    } catch (e: Exception) {
        val errorType = contentType.split("/").lastOrNull() ?: "UNKNOWN"
        sentryTrackMessage("NOTIFICATION_${errorType}_ERROR", mapOf("error" to e, "topic" to decodedMessage.topic))
        println("[NotificationExtension] ERROR WHILE HANDLING $contentType $e")
        return DecodedMessageResult(null, null, false, null)
    }
}

fun getContentTypeString(contentType: Content.ContentTypeId): String {
    return "${contentType.authorityId}/${contentType.typeId}:${contentType.versionMajor}.${contentType.versionMinor}"
}

fun saveMessageToStorage(appContext: Context, account: String, decodedMessage: DecodedMessage, content: String, contentType: String, referencedMessageId: String?) {
    val mmkv = getMmkv(appContext)
    val currentSavedMessagesString = mmkv?.decodeString("saved-notifications-messages")
    Log.d("PushNotificationsService", "Got current saved messages from storage: $currentSavedMessagesString")
    var currentSavedMessages = listOf<SavedNotificationMessage>()
    try {
        currentSavedMessages = Klaxon().parseArray<SavedNotificationMessage>(currentSavedMessagesString ?: "[]") ?: listOf()
    } catch (error: Exception) {
        Log.d("PushNotificationsService", "Could not parse saved messages from storage: $currentSavedMessagesString - $error")
    }
    if (content.isEmpty()) {
        return;
    }
    val newMessageToSave = SavedNotificationMessage(
        topic=decodedMessage.topic,
        content=content,
        senderAddress=decodedMessage.senderAddress,
        sent=decodedMessage.sent.time,
        id=decodedMessage.id,
        contentType=contentType,
        account=account,
        referencedMessageId=referencedMessageId
    )
    currentSavedMessages += newMessageToSave
    val newSavedMessagesString = Klaxon().toJsonString(currentSavedMessages)
    mmkv?.putString("saved-notifications-messages", newSavedMessagesString)
}

fun getJsonRemoteAttachment(remoteAttachment: RemoteAttachment): String {
    try {
        val dictionary =
            mapOf(
                "url" to remoteAttachment!!.url,
                "contentDigest" to remoteAttachment!!.contentDigest,
                "secret" to byteStringToBase64(remoteAttachment!!.secret),
                "salt" to byteStringToBase64(remoteAttachment!!.salt),
                "nonce" to byteStringToBase64(remoteAttachment!!.nonce),
                "scheme" to remoteAttachment!!.scheme,
                "contentLength" to (remoteAttachment!!.contentLength ?: 0),
                "filename" to (remoteAttachment!!.filename ?: "")
            )
        return JSONObject(dictionary).toString()
    } catch (e: Exception) {
        println("Error converting dictionary to JSON string: ${e.localizedMessage}")
        return "";
    }
}

fun getJsonReaction(decodedMessage: DecodedMessage): String {
    val reaction: Reaction? = decodedMessage.content()
    return try {
        val dictionary =
            mapOf(
                "action" to (reaction?.action?.javaClass?.simpleName?.lowercase() ?: ""),
                "schema" to (reaction?.schema?.javaClass?.simpleName?.lowercase() ?: ""),
                "content" to (reaction?.content ?: ""),
                "reference" to (reaction?.reference ?: ""),
            )
        JSONObject(dictionary).toString()
    } catch (e: Exception) {
        println("Error converting dictionary to JSON string: ${e.localizedMessage}")
        "";
    }
}


suspend fun handleV3Welcome(
    appContext: Context,
    xmtpClient: Client,
    conversation: Conversation,
    remoteMessage: RemoteMessage
): NotificationDataResult {
    var shouldShowNotification = false
    try {
        val mmkv = getMmkv(appContext)
        var apiURI = mmkv?.decodeString("api-uri")
        if (apiURI == null) {
            apiURI = getAsyncStorage("api-uri")
        }
        xmtpClient.syncConsent()
        val consentList = xmtpClient.preferences.consentList
        var spamScore = 1.0
        when (conversation) {
            is Conversation.Group -> {
                // Handle the Group case
                spamScore =
                    computeSpamScoreGroupWelcome(appContext, xmtpClient, conversation.group, apiURI)
            }

            is Conversation.Dm -> {
                // Handle the Dm case
                spamScore = computeSpamScoreDmWelcome(
                    appContext,
                    xmtpClient,
                    conversation.dm,
                    apiURI
                )
            }
        }
        if (spamScore < 0) { // Message is going to main inbox
            // consent list loaded in computeSpamScoreGroupWelcome
            val groupAllowed = isConversationAllowed(conversation, consentList)
            val groupDenied = isConversationBlocked(conversation, consentList)
            // If group is already consented (either way) then don't show a notification for welcome as this will likely be a second+ installation
            if (!groupAllowed && !groupDenied) {
                shouldShowNotification = true
            } else {
                shouldShowNotification = false
            }
        } else if (spamScore == 0.0) { // Message is Request
            shouldShowNotification = false
            // @todo : trackNewRequest()
        } else { // Message is Spam
            shouldShowNotification = false
        }
    } catch (e: Exception) {

    }
    conversation.sync()
    when (conversation) {
        is Conversation.Group -> {
            // Handle the Group case
            return NotificationDataResult(
                title = conversation.group.name,
                body = "You have been added to a new group",
                remoteMessage = remoteMessage,
                messageId = "welcome-${conversation.topic}",
                shouldShowNotification = shouldShowNotification
            )
        }
        is Conversation.Dm -> {
            // TODO:
            // Handle the Dm case
            return NotificationDataResult(
                title = conversation.dm.peerInboxId,
                body = "You have a new DM",
                remoteMessage = remoteMessage,
                messageId = "welcome-${conversation.topic}",
                shouldShowNotification = shouldShowNotification
            )
        }
    }

}

fun isGroupMessageFromMe(xmtpClient: Client, messageId: String): Boolean {
    try {
        val message = xmtpClient.findMessage(messageId)
        return message?.decodeOrNull()?.senderAddress == xmtpClient.inboxId;
    } catch (e: Exception) {
        return false
    }
}
